### DISCUSSION TITLE
Proper way to redirect on getServerSideProps? Error: Cannot set headers after they are sent.

### DISCUSSION THREAD

**Discussion by arielpacciaroni** [vercel/next.js#14664] [Help]
Hello! I'm using NextJS along Express and everytime I visit a protected page (through client-side navigation), I get the error described above "Cannot set headers after they are sent." I tried using res.end() and still gives me the error above. This is the snippet I'm using

```javascript
export const getServerSideProps = async ({ req, res }) => {
  const result = await Auth.isAuth(req);
  if (result.loggedIn === true) {
    return {
      props: {
        authUsername: result.username,
        loggedIn: true,
      },
    };
  } else {
    res.redirect("/signin");
    return { props: {} };
  }
};
```

Any help would be appreciated! Thanks!

**Comment by jamesmosier** [comment]
It looks like there was a Discussion on this [here](https://github.com/vercel/next.js/discussions/11281) with no resolution at this time. 

Some have suggested this (although I have not tested it):

```js
    ctx.res.setHeader('Location', '/my-page');
    ctx.res.statusCode = 302;
    ctx.res.end();
    return { props: {} };
```

**Reply by arielpacciaroni** [reply]
I still get the same error although it works as intended (same happens with res.redirect)

Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client

**Comment by AshConnolly** [comment]
I'd be interested in the solution to this.

I've also found this error in examples supplied by the next team, like the cool [static tweet demo](https://github.com/lfades/static-tweet) shared by @rauchg, code by @lfades, has this error every time it generates a new page. Would love to know what causes it! 👍

**Reply by lfades** [reply]
I think I saw it too, unfortunately I didn't have the time to go further into it and it is definitely a false positive 🤔

**Reply by AshConnolly** [reply]
Great to hear its a false positive, I'll ignore it for now, cheers! 😃
